spices/0002-uri-rewriting-for-mirroring.adoc

```diff
@@ -0,0 +1,319 @@
+= SPICE 2: URI Rewriting for Mirroring
+
+* Proposal: link:./0002-uri-rewriting-for-mirroring.adoc[SPICE-0002]
+* Author: link:https://github.com/holzensp[Philip Hölzenspies]
+* Status: Review
+* Implemented in: Pkl 0.26.0
+* Category: Tooling
+
+== Introduction
+
+This SPICE discusses rewrite rules for URIs of modules being `import`ed, so as to use mirrored alternatives.
+
+== Motivation
+
+For systems that require run-time air-gapping, packages imported from `pkg.pkl-lang.org` or other sites on the internet are not available.
+This can be remedied by pre-populating a cache directory with all required dependencies (see link:../spices/0001-import-graph-analyzer.pkl[SPICE 1: Import Graph Analyzer API]).
+When dependencies can not be determined ahead of time, pre-populating the cache is not an option.
+Mirroring packages from internet repositories in a local store is insufficient, because the mirrored packages may have hard-coded dependencies.
+Importing mirrored packages, using URIs pointing to the mirror, will eventually evaluate `import`s of those hard-coded dependencies, referring to external sites.
+
+Another reason to want to consume (transitive) dependencies from a mirror is the requirement of altered versions of packages.
+For example, redirecting a package developed for `github.com` to point to a GitHub Enterprise instance.
+Note that, because upstream dependencies specify the SHA256 hash, altering packages on a mirror requires _all upstream dependants to be updated_.
+This SPICE does _not_ provide further facilities for altering packages when mirroring.
+
+== Proposed Solution
+
+Similar to `git-config`'s `insteadOf`, the proposed solution is to rewrite URIs when used in `import`, `import*`, `amends`, `extends`, `read`, and `read*` clauses/expressions.
+From `git-config`'s `man` page, on `url.<base>.insteadOf`:
+
+> Any URL that starts with this value will be rewritten to start, instead, with `<base>`.
+> In cases where some site serves a large number of repositories, and serves them with multiple access methods, and some users need to use different access methods, this feature allows people to specify any of the equivalent URLs and have Git automatically rewrite the URL to the best alternative for the particular user, even for a never-before-seen repository on the site.
+> When more than one insteadOf strings match a given URL, the longest match is used.
+>
+> Note that any protocol restrictions will be applied to the rewritten URL.
+> If the rewrite changes the URL to use a custom protocol or remote helper, you may need to adjust the `protocol.*.allow` config to permit the request.
+
+The solution proposed in this SPICE mirrors this design, but extends it to allow regular expressions for the match.
+Only after rewriting are URIs tested for compliance with Allowed Modules and Allowed Resources settings.
+
+Rewrites can be provided:
+
+1. Via the API;
+2. Via `pkl:Project`; and
+3. Via arguments on the CLI.
```

Comment:
There's also:
* Gradle plugin
* Java evaluator API
* Server mode message passing API
* `~/.pkl/settings.pkl` file

In particular I think the settings file is an important place to be able to configure rewrite rules. Not everything needs to be a project, and in some cases can't be a project. This enables regular evaluation of Pkl without needing to introduce projects.

It also is really useful for, say, if you have a machine image (e.g. Docker) that you want to distribute that is already configured to talk to an internal host.

Reply:
"Via the API" was intended to refer to the Java Evaluator API (hence the proposed changes to `EvaluatorBuilder`). If this is reworded, are you expecting support for other APIs?

Similarly, how do we want to do these for SPICEs generally? List all surfaces where some work needs doing? Then it also needs to list every supported language binding / build system. I was generally assuming it understood that any change/addition of the Evaluator API _implies_ changes/additions to build systems / server / etc. Basically; do we want SPICEs to show the explicit design for _every_ contact surface?

Reply:
I think all places that are expected to require changes should at least be mentioned, especially those in apple/pkl. Supporting a new feature in the Evaluator API doesn’t necessarily mean that it should also be supported in the Gradle plugin.